framework/src/main/java/org/checkerframework/framework/ajava/TypeAnnotationMover.java

```diff
@@ -0,0 +1,226 @@
+        if (annoTypeElt == null) {
+            annoTypeElt = elements.getTypeElement(annoNameFq);
+            if (annoTypeElt == null) {
+                // Not a supported annotation.
+                return null;
+            }
+            AnnotationFileParser.putAllNew(
+                    allAnnotations,
+                    AnnotationFileParser.createNameToAnnotationMap(
+                            Collections.singletonList(annoTypeElt)));
+        }
+
+        return annoTypeElt;
+    }
+
+    /**
+     * Returns if {@code annotation} could be a declaration annotation for {@code declarationType}.
+     * This would be the case if the annotation isn't recognized at all, or if it was recognized and
+     * has {@code declarationType} as one of its targets.
+     *
+     * @param annotation a JavaParser annotation expression
+     * @param declarationType the declaration type to check if {@code annotation} might be a
+     *     declaration annotation for
+     * @return false unless {@code annotation} definitely cannot be a declaration annotation for
+     *     {@code declarationType}
+     */
+    private boolean isPossiblyDeclarationAnnotation(
+            AnnotationExpr annotation, ElementType declarationType) {
+        TypeElement annotationType = getAnnotationDeclaration(annotation);
+        if (annotationType == null) {
+            return true;
+        }
+
+        return isPossiblyDeclarationAnnotation(annotationType, declarationType);
+    }
+
+    /**
+     * Returns whether the annotation represented by {@code annotationDeclaration} might be a
+     * declaration annotation for {@code declarationType}. This holds if {@code declarationType} is
+     * a target of the annotation, or if {@code ElementType.TYPE_USE} is not a target of the
+     * annotation.
+     *
+     * @param annotationDeclaration declaration for an annotation
+     * @param declarationType the declaration type to check if the annotation might be a declaration
+     *     annotation for
+     * @return true if {@code annotationDeclaration} contains {@code declarationType} as a target or
+     *     doesn't contain {@code ElementType.TYPE_USE} as a target
+     */
+    private boolean isPossiblyDeclarationAnnotation(
+            TypeElement annotationDeclaration, ElementType declarationType) {
+        Target target = annotationDeclaration.getAnnotation(Target.class);
+        if (target == null) {
+            return true;
+        }
+
+        boolean hasTypeUse = false;
+        for (ElementType elementType : target.value()) {
+            if (elementType == ElementType.TYPE_USE) {
+                hasTypeUse = true;
+            }
+
+            if (elementType == declarationType) {
+                return true;
+            }
+        }
+
+        return !hasTypeUse;
+    }
+
+    /**
+     * Returns whether {@code type} has a name containing multiple parts separated by dots, e.g.
+     * "java.lang.String".
```

Comment:
I would also mention "Outer.Inner", because readers might not think of that case.  Does the code handle it correctly?